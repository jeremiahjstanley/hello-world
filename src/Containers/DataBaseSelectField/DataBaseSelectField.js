import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setDataBase } from '../../actions';
import { dataMetrics } from '../../helper/dataMetrics';

export class DataBaseSelectField extends Component {
  constructor() {
    super()

    this.state = {
      dataBase: {name: dataMetrics[0].name, database_code: dataMetrics[0].database_code}
    }
  }

  componentDidMount() {
    this.props.selectDataBase(this.state.dataBase)
  }

  selectMetric = (event) => {
    const dataBase = dataMetrics.find(metric => event.target.value === metric.name);
    const dataStore = {name: dataBase.name, database_code: dataBase.database_code}
    this.props.selectDataBase(dataStore)
  }

  render() {

    const options = dataMetrics.map((metric, index) => (
      <option key={index}>
        { metric.name }
      </option>
    ))

    return (
      <select onChange={this.selectMetric}>
        { options }
      </select>
    )
  };
}

export const mapDispatchToProps = (dispatch) => ({
  selectDataBase: (dataBase) => dispatch(setDataBase(dataBase))
})

export default connect(null, mapDispatchToProps)(DataBaseSelectField);