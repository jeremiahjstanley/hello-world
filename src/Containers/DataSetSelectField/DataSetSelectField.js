import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setDataSet } from '../../actions';
import { dataMetrics } from '../../helper/dataMetrics';

export class DataSetSelectField extends Component {
  constructor() {
    super()

    this.state = {
      dataSet: dataMetrics[0].datasets[0]
    }
  }

  componentDidMount() {
    this.props.selectDataSet(this.state.dataSet)
  }

  selectMetric = (event) => {
    const dataSet = this.props.dataSet.find(metric => event.target.value === metric.name);
    this.props.selectDataSet({...dataSet})
  }

  render() {
    // const options = this.props.dataSet.map((metric, index) => (
    //   <option key={index}>
    //     { metric.name }
    //   </option>
    // ))
    const options = 'cat'
    return (
      <select onChange={this.selectMetric}>
        { options }
      </select>
    )
  };
}

export const mapStateToProps = (state) => ({
  dataSet: state.dataBase.datasets 
})

export const mapDispatchToProps = (dispatch) => ({
  selectDataSet: (dataSet) => dispatch(setDataSet(dataSet))
})

export default connect(mapStateToProps, mapDispatchToProps)(DataSetSelectField);


