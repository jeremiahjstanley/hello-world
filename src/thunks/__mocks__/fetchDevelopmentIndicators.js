export const fetchDevelopmentIndicators = jest.fn().mockImplementation(() => {
  return {};
});