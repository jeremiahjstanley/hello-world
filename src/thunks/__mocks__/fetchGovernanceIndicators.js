export const fetchGovernanceIndicators = jest.fn().mockImplementation(() => {
  return {};
});