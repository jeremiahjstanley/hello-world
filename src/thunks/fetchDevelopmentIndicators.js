import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchDevelopmentIndicators = (locations, dataSet, dataBase) => {

    return async (dispatch) => {
        const unresolvedDevelopmentIndicators = locations.map(async location => {
            const isoAlpha3 = location.alpha_3
            try {
              const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}.json?api_key=${key}`;
              const response = await fetch(url);
              if (!response.ok) {
                throw Error(response.statusText);
              }
              return await response.json();
            } catch (error) {
              dispatch(hasErrored(true));
            }
        });

        const developmentIndicators = await Promise.all(unresolvedDevelopmentIndicators);
        const cleanDevelopmentIndicators = developmentIndicators.map(indicator => {
            return dataCleaner(indicator);
        });
        
        await dispatch(fetchLocationDataSuccess(cleanDevelopmentIndicators))   
    };
};