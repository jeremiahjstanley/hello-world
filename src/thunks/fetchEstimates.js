import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchEstimates = (isoAlpha3, dataSet, dataBase) => {
  return async (dispatch) => {
    try {
      const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}_EST.json?api_key=${key}`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return await response.json();
    } catch (error) {
      dispatch(hasErrored(true));
    }
  }
};