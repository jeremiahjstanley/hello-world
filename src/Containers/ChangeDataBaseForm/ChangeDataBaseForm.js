import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { fetchGovernanceIndicators } from '../../thunks/fetchGovernanceIndicators';
import { fetchDevelopmentIndicators } from '../../thunks/fetchDevelopmentIndicators';
import DataBaseSelectField from '../DataBaseSelectField/DataBaseSelectField';
import DataSetSelectField from '../DataSetSelectField/DataSetSelectField';

export class ChangeDataBaseForm extends Component {

  handleSubmit = (event) => {
    event.preventDefault();
    const { location, dataSet, dataBase} = this.props;
    ((dataBase === 'WWGI') ? 
      this.props.fetchGovernanceIndicators(location, dataSet, dataBase) : 
      this.props.fetchDevelopmentIndicators(location, dataSet, dataBase));
    this.props.history.push('/stats');
  };

  render() {

    return (

      <div className='change-form-container'>
        <Link to='/stats' className='back-button'>â—€ back</Link>
        <form onSubmit={this.handleSubmit} className='change-form'>
          <DataBaseSelectField/>
          <DataSetSelectField/>
          <button className='change-form-button'>Change</button>
       </form>
     </div>

    );
  };
};

export const mapStateToProps = (state) => ({
  dataBase: state.dataBase.database_code,
  dataSet: state.dataSet.dataset_code,
  location: state.location
});

export const mapDispatchToProps = (dispatch) => ({
  fetchDevelopmentIndicators: (locations, dataBase, dataSet) => dispatch(fetchDevelopmentIndicators(locations, dataBase, dataSet)),
  fetchGovernanceIndicators: (locations, dataBase, dataSet) => dispatch(fetchGovernanceIndicators(locations, dataBase, dataSet))

});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ChangeDataBaseForm));
