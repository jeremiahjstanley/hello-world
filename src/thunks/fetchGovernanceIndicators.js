import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { fetchEstimates } from './fetchEstimates';
import { fetchNumberOfSources } from './fetchNumberOfSources'; 
import { fetchPercentileRank } from './fetchPercentileRank';
import { fetchStandardError } from './fetchStandardError';

export const fetchGovernanceIndicators = (locations, dataSet, dataBase) => {
    return async (dispatch) => {
        const unresolvedGovernanceIndicators = locations.map(async location => {
            const alpha3 = location.alpha_3
            const estimates = await dispatch(fetchEstimates(alpha3, dataSet, dataBase));
            const numberOfSources = await dispatch(fetchNumberOfSources(alpha3, dataSet, dataBase)); 
            const percentileRank = await dispatch(fetchPercentileRank(alpha3, dataSet, dataBase));
            const standardError = await dispatch(fetchStandardError(alpha3, dataSet, dataBase));
            const cleanEstimates = dataCleaner(estimates);
            const cleanNumberOfSources = dataCleaner(numberOfSources);
            const cleanPercentileRank = dataCleaner(percentileRank);
            const cleanStandardError = dataCleaner(standardError);
            return {cleanEstimates, cleanNumberOfSources, cleanPercentileRank, cleanStandardError};
        });
        const governanceIndicators = await Promise.all(unresolvedGovernanceIndicators)
        await dispatch(fetchLocationDataSuccess(governanceIndicators))     
    };
};