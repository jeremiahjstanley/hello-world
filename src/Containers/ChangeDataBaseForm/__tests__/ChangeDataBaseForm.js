import React from 'react';
import { shallow } from 'enzyme'
import { ChangeDataBaseForm, mapStateToProps, mapDispatchToProps } from '../ChangeDataBaseForm';
import { fetchGovernanceIndicators } from '../../../thunks/fetchGovernanceIndicators';
import { fetchDevelopmentIndicators } from '../../../thunks/fetchDevelopmentIndicators';
import DataBaseSelectField from '../../DataBaseSelectField/DataBaseSelectField';
import DataSetSelectField from '../../DataSetSelectField/DataSetSelectField';

jest.mock('../../../thunks/fetchGovernanceIndicators');
jest.mock('../../../thunks/fetchDevelopmentIndicators');


describe('ChangeDataBaseForm', () => {

	let mockEvent
	let mockHistory;
  let mockLocation;
  let mockDataBase;
  let mockDataSet;
  let mockFetchDevelopmentIndicators;
  let mockFetchGovernanceIndicators;
  let wrapper;

  beforeEach(() => {

   	mockEvent = { preventDefault: jest.fn() };

  	mockHistory = [];

    mockLocation = [{
      name:'Burkina Faso',
      alpha_2:'BF',
      alpha_3:'BFA',
      country_code:'854',
      iso_3166_2:'ISO 3166-2: BF',
      region:'Africa',
      sub_region:'Sub-Saharan Africa',
      intermediate_region:'Western Africa',
      region_code:'002',sub_region_code:'202',
      intermediate_region_code:'011'
    }];

    mockDataBase = 'WWDI';

    mockDataSet = 'SM_POP_REFG';

	  mockFetchDevelopmentIndicators = jest.fn();

	  mockFetchGovernanceIndicators = jest.fn();

    wrapper = shallow(
    	<ChangeDataBaseForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		location={mockLocation}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		history={mockHistory}
    	/>
  	);

  });

	it('should match the snapshot', () => {

		expect(wrapper).toMatchSnapshot();

	});

 it('should invoke fetchGovernanceIndicators when the form is submitted with a database_code of WWDI', () => {

	  mockDataBase = 'WWGI'; 

    wrapper = shallow(
    	<ChangeDataBaseForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		location={mockLocation}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		history={mockHistory}
    	/>
    );

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchGovernanceIndicators).toHaveBeenCalledWith(mockLocation, mockDataSet, mockDataBase);

  });

   it('should invoke fetchDevelopmentIndicators when the form is submitted with a database_code other than WWGI', () => {

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchDevelopmentIndicators).toHaveBeenCalledWith(mockLocation, mockDataSet, mockDataBase);

  });

   it('should direct the user to the stats page of the application', () => {

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockHistory).toEqual(['/stats']);

  });

	describe('mapStateToProps', () => {

		let mockState;
		let mockDataBase;
		let mockDataSet;

		beforeEach(() => {

			mockDataBase = {
    		name: 'Worldwide Governance Indicators',
    		database_code: 'WWGI'
    	}

    	mockDataSet = {
        name: 'Voice and Accountability', 
        dataset_code: 'VA'
      }

			mockState = {
				dataBase: mockDataBase,
				dataSet: mockDataSet,
				location: mockLocation
			};

		});

		it('should return an object with dataBase, dataSet, and location key', () => {

			const expected = {
				dataBase: mockDataBase.database_code,
				dataSet: mockDataSet.dataset_code,
				location: mockLocation
			};

			const mappedProps = mapStateToProps(mockState);

			expect(mappedProps).toEqual(expected);

		});

	});

	describe('mapDispatchToProps', () => {

		let actionToDispatch;
		let mockDispatch;

		beforeEach(() => {

			mockDispatch = jest.fn();

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

	  	mockDataBase = 'WWGI'; 

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});
		
	});

});