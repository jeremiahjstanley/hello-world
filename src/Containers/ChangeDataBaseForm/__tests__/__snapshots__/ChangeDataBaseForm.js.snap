// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeDataBaseForm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChangeDataBaseForm
    dataBase="WWDI"
    dataSet="SM_POP_REFG"
    fetchDevelopmentIndicators={[Function]}
    fetchGovernanceIndicators={[Function]}
    history={Array []}
    location={
        Array [
            Object {
              "alpha_2": "BF",
              "alpha_3": "BFA",
              "country_code": "854",
              "intermediate_region": "Western Africa",
              "intermediate_region_code": "011",
              "iso_3166_2": "ISO 3166-2: BF",
              "name": "Burkina Faso",
              "region": "Africa",
              "region_code": "002",
              "sub_region": "Sub-Saharan Africa",
              "sub_region_code": "202",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          className="back-button"
          replace={false}
          to="/stats"
>
          ◀ 
          <span>
                    back
          </span>
</Link>,
        <form
          className="change-form"
          onSubmit={[Function]}
>
          <Connect(DataBaseSelectField) />
          <Connect(DataSetSelectField) />
          <button
                    className="change-form-button"
          >
                    Change
          </button>
</form>,
      ],
      "className": "change-form-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "◀ ",
            <span>
              back
</span>,
          ],
          "className": "back-button",
          "replace": false,
          "to": "/stats",
        },
        "ref": null,
        "rendered": Array [
          "◀ ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "back",
            },
            "ref": null,
            "rendered": "back",
            "type": "span",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Connect(DataBaseSelectField) />,
            <Connect(DataSetSelectField) />,
            <button
              className="change-form-button"
>
              Change
</button>,
          ],
          "className": "change-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Change",
              "className": "change-form-button",
            },
            "ref": null,
            "rendered": "Change",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            className="back-button"
            replace={false}
            to="/stats"
>
            ◀ 
            <span>
                        back
            </span>
</Link>,
          <form
            className="change-form"
            onSubmit={[Function]}
>
            <Connect(DataBaseSelectField) />
            <Connect(DataSetSelectField) />
            <button
                        className="change-form-button"
            >
                        Change
            </button>
</form>,
        ],
        "className": "change-form-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "◀ ",
              <span>
                back
</span>,
            ],
            "className": "back-button",
            "replace": false,
            "to": "/stats",
          },
          "ref": null,
          "rendered": Array [
            "◀ ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "back",
              },
              "ref": null,
              "rendered": "back",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Connect(DataBaseSelectField) />,
              <Connect(DataSetSelectField) />,
              <button
                className="change-form-button"
>
                Change
</button>,
            ],
            "className": "change-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Change",
                "className": "change-form-button",
              },
              "ref": null,
              "rendered": "Change",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
